### ISSUE TITLE
asyncValidate doesn't allow objects for errors

### ISSUE THREAD

**Issue by mlarcher** [erikras/redux-form#506]
With the `validate` function we can return Objects with error name and parameters, which come in very handy for error reporting. However, if we do the same from an `asyncValidate`function, the object gets lost in the way and the error doesn't appear at all on the field object.
Is this a bug or is it intentional ?

**Comment by erikras** [comment]
Sounds like a bug. Can you give an example of a value returned from `asyncValidate` that fails in this way?

**Comment by mlarcher** [comment]
I have the problem with the following object as error:
```
return {
  name: "SOME_ERROR_CONSTANT"
};
```
if I just use
```
return "SOME_ERROR_CONSTANT";
```
instead, I get the expected error.

**Comment by erikras** [comment]
That's very odd. Have another look at [the example](http://erikras.github.io/redux-form/#/examples/asynchronous-blur-validation), and see if you can notice anything you're doing differently.

**Comment by mlarcher** [comment]
I had hard time figuring out what was going wrong, and got to the example many times. In the end, I only had something like this:
```
export default reduxForm({
  form: 'myForm',
  fields: ['username'],
  asyncValidate: (values/*, dispatch */) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
        reject({username: {
            name: 'That username is taken'
        }
        });
    }, 100);
  });,
  asyncBlurFields: ['username']
})(MyComponent);
```
which was ignoring the error, while
```
export default reduxForm({
  form: 'myForm',
  fields: ['username'],
  asyncValidate: (values/*, dispatch */) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
        reject({username:  'That username is taken'
        });
    }, 100);
  });,
  asyncBlurFields: ['username']
})(MyComponent);
```
was displaying it.

I tried to follow the calls in the lib, going from asyncValidation.js _isValid2 to isValid, which seemed like an odd function to apply to an error output (as opposed to a field or field value). I didn't figure out whaat was going on in there with the recursive calls and stopped my investigation there.

**Comment by erikras** [comment]
What is this `name` key for? Your data looks like this:
```js
{
  username: 'mlarcher'
}
```
So your error needs the same shape:
```js
{
  username: 'invalid for some reason'
}
```
If you give an error like:
```js
{
  username: {
    name: 'some error'
  }
}
```
...redux-form is going to assume that your data looks like:
```js
{
  username: {
    name: 'mlarcher'
  }
}
```
Perhaps redux-form should be smarter, and look at your `fields` array to see the shape of your data and accept objects as errors, but, at the moment, it does not.

Does all that make sense?

**Comment by mlarcher** [comment]
I understand. I implemented the synchronous validation first and didn't face this problem, so I was a bit surprised when it arose during async validation implementation.
A string is not enough for us as an error content, as we need to be able to pass params along with the error key for better error reporting. 
What approach would you recommend to end up with an object as the error content ?

**Comment by erikras** [comment]
Ah. This is a bug, then. Async and submit errors are saved in the state, but sync errors are not, so there must be some inconsistency there. Investigating... :mag:

**Comment by erikras** [comment]
Fix published in [`v4.1.1`](https://github.com/erikras/redux-form/releases/tag/v4.1.1).

Thanks for this.

**Comment by mlarcher** [comment]
Great! Thanks for the quick fix :)
